lib/bats-core/test_functions.bash

```diff
@@ -3,30 +3,179 @@
+  IFS=: read -ra bats_lib_paths <<< "$bats_lib_path"
+
+  for path in "${bats_lib_paths[@]}"; do
+    if find_library_load_path "$path/$library_name"; then
+      # A library load path was found, return
+      return
+    fi
+  done
+
+  return 1
+}
+
+# _load expects an absolute path that is a library load path.
+#
+# If the library load path points to a file (a library loader) it is
+# sourced.
+#
+# If it points to a directory all files ending in .bash inside of the
+# directory are sourced.
+#
+# If the sourcing of the library loader or of a file in a library
+# directory fails _load prints an error message and returns 1.
+#
+# If the passed library load path is not absolute or is not a valid file
+# or directory _load prints an error message and returns 1.
+_load() {
+  local library_load_path="${1:?}"
+
+  if [[ "${library_load_path:0:1}" != / ]]; then
+    printf "Passed library load path is not an absolute path: %s\n" "$library_load_path" >&2
+    return 1
+  fi
+
+  # library_load_path is a library loader
+  if [[ -f "$library_load_path" ]]; then
+      if ! source "$library_load_path"; then
+          printf "Error while sourcing library loader at '%s'\n" "$library_load_path" >&2
+          return 1
+      fi
+      return
+  fi
+
+  # library_load_path is a library directory
+  if [[ -d "$library_load_path" ]]; then
+    for library_file in "$library_load_path"/*.bash; do
 
-  # Dynamically loaded user file provided outside of Bats.
-  # Note: 'source "$file" || exit' doesn't work on bash3.2.
-  # shellcheck disable=SC1090
-  source "${file}"
+      # Skip over directories
+      [[ -d "$library_file" ]] && continue
+
+      if ! source "$library_file"; then
+        printf "Error while sourcing library file '%s'\n" "$library_file" >&2
+        return 1
+      fi
+    done
+
+    return
+  fi
+
+  printf "Passed library load path is neither a library loader nor library directory: %s\n" "$library_load_path" >&2
+  return 1
+}
+
+# load_safe accepts an argument called 'slug' and attempts to find and
+# source a library based on the slug.
+#
+# A slug can be an absolute path, a library name or a relative path.
+#
+# If the slug is an absolute path load_safe attempts to find the library
+# load path using find_library_load_path.
+# What is considered a library load path is documented in the
+# documentation for find_library_load_path.
+#
+# If the slug is not an absolute path it is considered a library name or
+# relative path. load_safe attempts to find the library load path using
+# find_in_bats_lib_path.
+#
+# If load_safe can find a library load path it is passed to _load.
+# If _load fails load_safe returns 1.
+#
+# If no library load path can be found load_safe prints an error message
+# and returns 1.
+load_safe() {
+  local slug="${1:?}"
+
+  # Check if slug is an absolute path
+  if [[ "${slug:0:1}" == / ]]; then
+
+    # Check for library load paths
+    local library_path="$(find_library_load_path "$slug")"
```

Comment:
Won't `find_library_load_path` try to prepend prefixes instead of using this absolute path as is?

Maybe we should have additional tests for that.

Reply:
I mistook that for `find_in_bats_lib_path`. `find_library_load_path` just tries the common extensions as expected. Also, there is already a test ("load supports plain symbols") that loads absolute paths.